src/main/resources/META-INF/rewrite/micronaut3-to-4.yml

```diff
@@ -0,0 +1,373 @@
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.graalvm
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.library
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.minimal.application
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.minimal.library
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.test-resources
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.test-resources-consumer
+      newVersion: 4.0.0-M2
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateMicronautPlatformBom
+displayName: Update to Micronaut 4.x platform BOM
+description: This recipe will update a Gradle or Maven build to reference the Micronaut 4 platform BOM.
+recipeList:
+  - org.openrewrite.maven.ChangeParentPom:
+      oldGroupId: io.micronaut
+      oldArtifactId: micronaut-parent
+      newGroupId: io.micronaut.platform
+      newVersion: 4.0.0-M2
+      retainVersions:
+        - io.micronaut:micronaut-validation
+        - io.micronaut:micronaut-session
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateJakartaAnnotations
+displayName: Update jakarta annotations dependency
+description: This recipe will remove jakarta annotations dependency as it is a transitive dependency of micronaut-inject, and migrate from javax.annotation if needed.
+recipeList:
+  - org.openrewrite.java.migrate.jakarta.JavaxAnnotationPackageToJakarta
+  - org.openrewrite.gradle.RemoveDependency:
+      groupId: jakarta.annotation
+      artifactId: jakarta.annotation-api
+  - org.openrewrite.maven.RemoveDependency:
+      groupId: jakarta.annotation
+      artifactId: jakarta.annotation-api
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateJakartaPersistence
+displayName: Update jakarta persistence
+description: This recipe will migrate from javax.persistence annotations if needed.
+recipeList:
+  - org.openrewrite.java.ChangePackage:
+      oldPackageName: javax.persistence
+      newPackageName: jakarta.persistence
+      recursive: true
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateMicronautValidation
+displayName: Update to Micronaut Validation 4.x
+description: This recipe will add jakarta validation dependency if needed, migrate from javax.validation if needed, and update micronaut validation dependencies.
+recipeList:
+  - org.openrewrite.java.ChangePackage:
+      oldPackageName: javax.validation
+      newPackageName: jakarta.validation
+      recursive: true
```

Comment:
And there's also a `org.openrewrite.java.migrate.jakarta.JavaxValidationMigrationToJakartaValidation`